lib/Solver/Solver.cpp

```diff
@@ -0,0 +1,363 @@
+    if (!llvm::dyn_cast<vara::feature::BinaryFeature>(&FeatureToAdd)) {
+      return NOT_SUPPORTED;
+    }
+    auto R = setBinaryFeatureConstraints(
+        *llvm::dyn_cast<vara::feature::BinaryFeature>(&FeatureToAdd),
+        IsInAlternativeGroup);
+    if (!R) {
+      return R;
+    }
+    break;
+  }
+  case feature::Feature::FeatureKind::FK_ROOT:
+    addFeature(FeatureToAdd.getName().str());
+    // Add root as a constraint; root is mandatory
+    Solver->add(*OptionToVariableMapping[FeatureToAdd.getName()]);
+    break;
+  case feature::Feature::FeatureKind::FK_UNKNOWN:
+    return NOT_SUPPORTED;
+  }
+
+  Solver->push();
+  return Ok();
+}
+
+Result<SolverErrorCode> Z3Solver::addFeature(const string &FeatureName) {
+  if (OptionToVariableMapping.find(FeatureName) !=
+      OptionToVariableMapping.end()) {
+    return ALREADY_PRESENT;
+  }
+  z3::expr Feature = Context.bool_const(FeatureName.c_str());
+  OptionToVariableMapping.insert(
+      std::make_pair(FeatureName, std::make_unique<z3::expr>(Feature)));
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::addFeature(const string &FeatureName,
+                     const std::vector<int64_t> &Values) {
+  if (OptionToVariableMapping.find(llvm::StringRef(FeatureName)) !=
+      OptionToVariableMapping.end()) {
+    return ALREADY_PRESENT;
+  }
+  z3::expr Feature = Context.int_const(FeatureName.c_str());
+  OptionToVariableMapping.insert(
+      std::make_pair(FeatureName, std::make_unique<z3::expr>(Feature)));
+
+  // Add the numeric values as constraints
+  z3::expr Constraint = Context.bool_val(false);
+  for (int64_t Value : Values) {
+    Constraint = Constraint || (Feature == Context.int_val(Value));
+  }
+  Solver->add(Constraint);
+
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::removeFeature(feature::Feature &FeatureToRemove) {
+  return NOT_SUPPORTED;
+}
+
+Result<SolverErrorCode>
+Z3Solver::addRelationship(const feature::Relationship &R) {
+  const auto *Parent = (const feature::Feature *)R.getParent();
+  auto ParentOption = Context.bool_const(Parent->getName().str().c_str());
+  z3::expr_vector V(Context);
+  for (const auto &Child : R.children()) {
+    const auto *ChildFeature = (const feature::Feature *)Child;
+    V.push_back(Context.bool_const(ChildFeature->getName().str().c_str()));
+  }
+  switch (R.getKind()) {
+  case feature::Relationship::RelationshipKind::RK_ALTERNATIVE:
+    Solver->add(z3::implies(ParentOption, z3::atmost(V, 1) & z3::mk_or(V)));
+    break;
+  case feature::Relationship::RelationshipKind::RK_OR:
+    Solver->add(z3::implies(ParentOption, z3::mk_or(V)));
+    break;
+  }
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::addConstraint(feature::Constraint &ConstraintToAdd) {
+  SolverConstraintVisitor SCV(this);
+  bool Succ = SCV.addConstraint(&ConstraintToAdd);
+  if (!Succ) {
+    return SolverErrorCode::NOT_SUPPORTED;
+  }
+  return Ok();
+}
+
+Result<SolverErrorCode, std::unique_ptr<bool>>
```

Comment:
Why is the bool here store in a unique_ptr and not just copied?

Reply:
That's right. That's a bit too much. I will change that.